lib/AST/TS2Flow.cpp

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "hermes/AST/TS2Flow.h"
+
+#include "hermes/AST/RecursiveVisitor.h"
+
+namespace hermes {
+
+namespace {
+
+class TS2FlowValidator {
+ public:
+  explicit TS2FlowValidator(Context &context) : context_(context) {}
+
+  /// Check if the given program has any TS nodes.
+  /// \param program the top-level program node.
+  /// \return true if the program has no TS node, false otherwise.
+  bool validate(ESTree::ProgramNode *program) {
+    auto &sm = context_.getSourceErrorManager();
+    if (sm.getErrorCount())
+      return false;
+
+    visitESTreeNode(*this, program);
+
+    return sm.getErrorCount() == 0;
+  }
+
+  bool incRecursionDepth(ESTree::Node *) {
+    return true;
+  }
+
+  void decRecursionDepth() {}
+
+  void visit(ESTree::Node *node) {
+    auto kind = node->getKind();
+
+    // TSTypeAnnotation is the first TS node and TSModifiers is the last.
+    // This must be in sync with ESTree.def.
```

Comment:
These checks that have to be kept in sync with ESTree.def seem fragile, and I've wanted this functionality before. 

I'm going to go ahead and make all the Flow nodes inherit from a FlowNode class and the TS nodes inherit from a TS class, so you can just do `llvh::isa<TSNode>(node)`, e.g. I'll try and get that done ASAP.

Reply:
@avp's commit for this already landed: https://github.com/facebook/hermes/commit/0d2003c8af30ec258663ed0344f54a97ba061c40